### ISSUE TITLE
Inference doesn't finish

### ISSUE THREAD

**Issue by KnutJaegersberg** [mobiusml/hqq#4]
Hello, I wanted to 2-bit quantize a couple of long context llama based models. 
I did so for causallm-72b-35k, yi-34b-200k, giraffe-70b-32k, all with group size 16.  
I've uploaded causallm:
https://huggingface.co/KnutJaegersberg/CausalLM-72B-preview-hqq-2bit

The problem I face is when I use the example code, inference never finishes. My GPU get's busy, but nothing else happens. All models I did have this. I should have tried first, but instead I just executed the quantization code several times. 
Are these quantizations now useless?

**Comment by mobicham** [comment]
@KnutJaegersberg have you tried the same models with a shorter context size? With a long context-size, the speed should be actually closer to fp16. My initial guess is that torch compile is taking a long time. Try a warm-up like this before running the generation: 

```
x = torch.ones([1, context_length], dtype=torch.int32, device='cuda')
for _ in range(5):
	with torch.no_grad(): _ = model(x)
	print('...')
```

Can you share your quantization code as well ?

**Comment by KnutJaegersberg** [comment]
Thanks, I'll try that as soon my GPU finished a long running job (2 days or so). 
I used the code from your example, I think all I changed was group size. 
Didn't have enough ram so I made my swap file larger, but I would not expect that to have any impact. 

from hqq.engine.hf import HQQModelForCausalLM, AutoTokenizer
model_id   = 'path to local copy of causalLM'

#Load model on the CPU
######################
model     = HQQModelForCausalLM.from_pretrained(model_id)
tokenizer = AutoTokenizer.from_pretrained(model_id) 

#Quantize the model
######################
from hqq.core.quantize import *
quant_config = BaseQuantizeConfig(nbits=2, group_size=16)
model.quantize_model(quant_config=quant_config)

**Comment by mobicham** [comment]
Thanks for the code snippet! So you didn't implement a custom quantization for your models. Currently, we only support Llama2 and Mixtral. If these models are not based on Llama2 or Mixtral, you'd have to create your own. But the fact that it was able to quantize it means that they are Llama-based? Please double check the architecture and see if the layers are named the same way as the Llama2 models.

Otherwise it's easy to do, for Hugging Face models: 
- Go to https://github.com/mobiusml/hqq/tree/master/hqq/models/hf 
- Copy the llama.py to yi.py for example, customize it for your yi model (change LlamaHQQ to the class of your choice, like YiHQQ for example). Just change the names of the layers
- then use it like this:
```python 
from hqq.models.hf.yi import YiHQQ
YiHQQ.quantize_model(model, quant_config=quant_config)
YiHQQ.save_quantized(model=model, save_dir=save_dir)
#Don't forget to save the tokenizer in the same directory
tokenizer.save_quantized(save_dir)
```
```python 
#Load
model     = YiHQQ.from_quantized(save_dir_or_hf_hub)
tokenizer = AutoTokenizer.from_pretrained(save_dir_or_hf_hub)
```
Do this for each family of models. 

By the way, for 2-bit you should use this instead:
`quant_config = BaseQuantizeConfig(nbits=2, group_size=16, quant_zero=True, quant_scale=True)`

**Comment by KnutJaegersberg** [comment]
Thanks. I think the models I tried are all llama architecture based, though not official llamas. 
What is the impact of doing 2 bit without the last two arguments?

**Comment by KnutJaegersberg** [comment]
I think the yi models have been llamafied somewhat after their first release. causallm should be llama.

**Comment by mobicham** [comment]
Looking at the Yi config, it is indeed Llama, so it should work fine. But try to avoid making the first few calls with a large context-size, so torch.compile doesn't get stuck compiling. Actually, try with `HQQLinear.set_backend(HQQBackend.PYTORCH) ` first.

Regarding the 2-bit parameters, the `quant_zero=True` is actually by default, but adding `quant_scale=True` is important for low group-sizes like 16 otherwise the size of the quantized model will be a bit larger. Basically it's quantizing the scaling/zero-point parameters to 8-bit which makes the model smaller with a negligible accuracy loss.
